packages/core-data/src/footnotes/index.js

```diff
@@ -0,0 +1,110 @@
+/**
+ * Internal dependencies
+ */
+import getFootnotesOrder from './get-footnotes-order';
+
+let oldFootnotes = {};
+
+export function updateFootnotesFromMeta( blocks, meta ) {
+	const output = { blocks };
+	if ( ! meta ) return output;
+
+	// If meta.footnotes is empty, it means the meta is not registered.
+	if ( meta.footnotes === undefined ) return output;
+
+	const newOrder = getFootnotesOrder( blocks );
+
+	const footnotes = meta.footnotes ? JSON.parse( meta.footnotes ) : [];
+	const currentOrder = footnotes.map( ( fn ) => fn.id );
+
+	if ( currentOrder.join( '' ) === newOrder.join( '' ) ) return output;
+
+	const newFootnotes = newOrder.map(
+		( fnId ) =>
+			footnotes.find( ( fn ) => fn.id === fnId ) ||
+			oldFootnotes[ fnId ] || {
+				id: fnId,
+				content: '',
+			}
+	);
+
+	function updateAttributes( attributes ) {
+		attributes = { ...attributes };
+
+		for ( const key in attributes ) {
+			const value = attributes[ key ];
+
+			if ( Array.isArray( value ) ) {
+				attributes[ key ] = value.map( updateAttributes );
+				continue;
+			}
+
+			if ( typeof value !== 'string' ) {
+				continue;
+			}
+
+			if ( value.indexOf( 'data-fn' ) === -1 ) {
+				continue;
+			}
+
+			// When we store rich text values, this would no longer
+			// require a regex.
```

Comment:
no regex is required today, no? and it's JS so it's easy to build a DOM of the HTML and then modify it without risking more corruption.

it's also unclear to me what our purpose is here, partially because of the obscurity of the regex patterns themselves. `regex` is moving the footnote id into the link text?

```js
const dom = parseHTML( value );

// Set the footnote id as the link's text.
dom.querySelectorAll( 'sup[data-fn] > a' ).forEach( ( node ) => {
	const fnId = node.parentNode.getAttribute( 'data-fn' );

	node.innerText = newOrder.indexOf( fnId ) + 1;
} );

// Replace footnotes that are only anchor tags with the <sup>-wrapped version.
// From: <a data-fn="a431c">a431c</a>
//   To: <sup data-fn="a431c" class="fn"><a href="#a431c" id="a431c-link">a431c</a></sup>
dom.querySelectorAll( 'a[data-fn]' ).forEach( ( node ) => {
	const fnId = node.parentNode.getAttribute( 'data-fn' );

	const sup = dom.createElement( 'sup' );
	sup.setAttribute( 'data-fn', fnId );
	sup.classList.add( 'fn' );

	const link = dom.createElement( 'a' );
	link.setAttribute( 'href', `#${ fnId }` );
	link.setAttribute( 'id', `${ fnId }-link` );
	link.innerText = fnId + 1;

	sup.appendChild( link );
	node.parentNode.replaceNode( node, sup );
} );
```

Reply:
These regexes are just temporary until we get rich text values in the store. If we can ever do that at least :(
We could cache the rich text values instead somewhere, but we can't use a weak map since the key would be a string... Any ideas?

I would prefer if we could cache the rich text values themselves somewhere instead of caching rich text values for a block.